Fix max_keys reset limitation

theamirocohen opened the pull request:
Fix issue NvStore does not correctly persist values through reset
Resolves [#8785](https://github.com/ARMmbed/mbed-os/issues/8785)
The former NvStore implementation did not persist the max_keys value through a soft-reset, a set value would have disappear after reset, meaning NvStore would start over with its default max_keys value which may cause data to be lost.
We added the max_keys value to master record (written to non volatile area) so we can read and update its value after reset.
Furthermore we added the restriction of setting max_keys to be bigger than the default predefined keys value (16), and in case we have set max_keys to a new value and want to set it again to a smaller value the operation will fail if there are keys that might be lost (if keys were store at that gap).

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

ciarmcom commented:
@theamirocohen, thank you for your changes.
@ARMmbed/mbed-os-storage @ARMmbed/mbed-os-maintainers please review.

yossi2le reviewed (approved):
LGTM

0xc0170 commented:
@theamirocohen Can you add more details to the commit msg ? The fix does not look that trivial.

cc @NeilMacMullen for review

0xc0170 commented:
I set this to 5.12 but I noticed this is updating private method, so no functional change there (fix can go to the next patch) ?

theamirocohen commented:
> I set this to 5.12 but I noticed this is updating private method, so no functional change there (fix can go to the next patch) ?

Yes, fix can go to next patch

theamirocohen committed:
Fix max_keys reset limitation

Persist the max_keys value through a soft-reset, also prohibit max_keys set under predefined default value (16)

cmonr commented:
CI started

cmonr commented:
Interesting. Error indicates that the NRF52_DK failed to flash.

CI job restarted: `jenkins-ci/greentea-test`
